File: frontend/src/api/mappers/hotelMapper.ts
import type { HotelDTO, RoomDTO, BookingDTO } from "../dto";
import type { IHotel, IRoom, IBooking } from "../domain/models";

export class HotelMapper {
	static toDomain(dto: HotelDTO): IHotel {
		return {
			id: dto.id,
			name: dto.name,
			location: dto.address,
			rooms: dto.rooms.map((r) => this.mapRoom(r)),
		};
	}

	private static mapRoom(dto: RoomDTO): IRoom {
		return {
			id: dto.id,
			title: `Номер ${dto.number}`,
			cost: `$${dto.price}`,
			image: dto.imageUrl,
			bookings: dto.bookings.map((b) => this.mapBooking(b)),
		};
	}

	private static mapBooking(dto: BookingDTO): IBooking {
		return {
			id: dto.id,
			period: `${dto.checkIn} — ${dto.checkOut}`,
			guest: dto.guestName,
			isActive: dto.status === "CONFIRMED",
			status: dto.status,
		};
	}
}

File: mobile/lib/mappers/hotel_mapper.dart
import '../core/dtos.dart';
import '../domain/models.dart';

class HotelMapper {
  static Hotel toDomain(HotelDto dto) {
    return Hotel(
      id: dto.id,
      name: dto.name,
      location: dto.address,
      rooms: dto.rooms.map(_mapRoom).toList(),
    );
  }

  static Room _mapRoom(RoomDto dto) {
    return Room(
      id: dto.id,
      title: 'Номер ${dto.number}',
      formattedPrice: '\$${dto.price.toStringAsFixed(0)}',
      imageUrl: dto.imageUrl,
      bookings: dto.bookings.map(_mapBooking).toList(),
    );
  }

  static Booking _mapBooking(BookingDto dto) {
    return Booking(
      id: dto.id,
      period: '${dto.checkIn} — ${dto.checkOut}',
      guest: dto.guestName,
      isActive: dto.status == 'CONFIRMED',
    );
  }

  static Room toDomainRoom(RoomDto dto) {
    return _mapRoom(dto);
  }
}
